baselines/jft/al_utils.py

```diff
@@ -12,81 +12,134 @@
-
-"""Custom CIFAR-10 dataset that returns a subset for training.
-
-Alternative version that writes out TFRecords here:
-https://colab.research.google.com/drive/1McRC0es1ehwUL_jQQcBdC05wsPGgNWE3
-"""
+import dataclasses
 import logging
+import re
+
 import jax
+
+from typing import Dict, Union, Optional, List, Set, Iterable
+
 import numpy as np
 import tensorflow as tf
+from clu.deterministic_data import DatasetBuilder
 import tensorflow_datasets as tfds
 
-cifar = tfds.image_classification.cifar
 
+# adapted from https://www.tensorflow.org/datasets/determinism
+class TFDSIdToInt:
+  tfds_id_parser_re = re.compile(r'\w+-(\w+).\w+-(\d+)-of-\d+__(\d+)')
+
+  shard_offsets: Dict[str, List[int]]
+
+  def __init__(self, dataset_info: tfds.core.DatasetInfo, *,
+               splitwise_id: bool):
+    shard_offsets = {}
+    current_offset = 0
+
+    # Build a unique offset across splits or within splits depending on `splitwise_id`.
+    for split, split_info in sorted(dataset_info.splits.items()):
+      if splitwise_id:
+        current_offset = 0
+
+      split_shard_offsets = []
+      for shard_length in split_info.shard_lengths:
+        current_length = current_offset + shard_length
+        split_shard_offsets.append(current_offset)
+        current_offset = current_length
+
+      shard_offsets[split] = split_shard_offsets
 
-def _subset_generator(dataset, subset_ids):
+    self.shard_offsets = shard_offsets
+
+  def __call__(self, tfds_id: str) -> int:
+    """Format the tfds_id in a more human-readable."""
+    match = self.tfds_id_parser_re.match(tfds_id)
+    split_name, shard_id, ex_id = match.groups()
+    shard_offsets = self.shard_offsets[split_name]
+    return shard_offsets[int(shard_id)] + int(ex_id)
+
+
+def _subset_generator(*, dataset: tf.data.Dataset,
+                      dataset_info: tfds.core.DatasetInfo,
+                      subset_ids: Optional[Set[int]], splitwise_id: bool):
   """Create a subset based on ids."""
+  tfds_id_to_int = TFDSIdToInt(dataset_info, splitwise_id=splitwise_id)
 
   def inner():
-    for record in dataset:
-      # HACK: the last 5 characters of 'id' are generally the ids.
-      # However, the dataset info only specifies that the type of id
-      # is BYTES, so we also support that.
-      try:
-        int_id = np.int32(record['id'].numpy()[-5:])
-      except ValueError:
-        # ID is encoded differently, then just interpret as bytes
-        int_id = int.from_bytes(record['id'].numpy(), 'big')
-        # jax prefers int32
-        # https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision
-        int_id = np.int32(int_id % np.iinfo(np.int32).max)
-
-      if subset_ids is not None:
-        # Hard fail on type errors
-        assert all(map(lambda id: isinstance(id, int), subset_ids))
+    if subset_ids is not None:
+      # Hard fail on type errors
+      assert all(map(lambda id: isinstance(id, int), subset_ids))
```

Comment:
This is probably a bit too defensive, we would notice it if it wasn't true.

Reply:
I think it's good to fail really hard here because there are multiple code paths downstream and if a few non-ints snuck through it would be near impossible to find out otherwise.

Likewise, given that we don't expect to collect more than 200-300 samples, the runtime cost is manageable.

An alternative would be to add an 'acquire' method that encapsulates subset_ids away and have the check there only once.